### PR TITLE
vhost-user-net: Allow reuse of existing tap interface

### LINKED ISSUE #742: vhost-user-net backend needs to support using existing tap interface
And we need to do the same checks that the virtio-net backend does: #741

### PR DESCRIPTION
**Pull request by bjzhjing**
Add tap option for vhost-user-net backend, it will allow to use the existing Tap interface.

Fix #742

### CODE REVIEW

**Comment by rbradford** [comment]
lgtm. Integration test needed that uses an existing interface though.

**Comment by bjzhjing** [comment]
> lgtm. Integration test needed that uses an existing interface though.

make use of an existing interface in test.

File: scripts/run_integration_tests.sh [hunk 1]
```diff
@@ -155,6 +155,9 @@ sudo ip tuntap add vfio-tap2 mode tap
 sudo ip link set vfio-tap2 master vfio-br0
 sudo ip link set vfio-tap2 up
 
+# Create tap interface with multipe queue support for vhost_user_net test.
```
Comment (sboeuf) [inline scripts/run_integration_tests.sh]: typo: s/multipe/multiple/

File: tests/integration.rs [hunk 2]
```diff
@@ -1113,13 +1113,14 @@ mod tests {
...
+            let vunet_tap0 = "vunet-tap0";
+
             // Start the daemon
             let mut daemon_child = Command::new("target/release/cloud-hypervisor")
                 .args(&[
                     "--net-backend",
                     format!(
-                        "ip={},mask=255.255.255.0,sock=/tmp/vunet.sock,num_queues=4,queue_size=1024",
-                        guest.network.host_ip
+                        "tap={},ip={},mask=255.255.255.0,sock=/tmp/vunet.sock,num_queues=4,queue_size=1024", vunet_tap0, guest.network.host_ip
```
Comment (sboeuf) [inline tests/integration.rs]: The problem here is that you modify an existing test, which modifies the previous behavior. Are we okay about testing only when the tap interface is already created? Or should we test both use cases?

Comment (rbradford) [reply tests/integration.rs]: Good catch. We need to test both. There are a lot of variables here so it is probably a good idea to factorise the test like the virtio-fs ones

File: tests/integration.rs [hunk 3]
```diff
@@ -410,6 +410,46 @@ mod tests {
...
+                    "--net-backend",
+                    format!(
+                        "ip={},mask=255.255.255.0,sock={},num_queues={},queue_size=1024",
+                        ip, vunet_socket_path, num_queues
+                    )
+                    .as_str(),
+                ])
+                .spawn()
+                .unwrap()
+        };
```
Comment (sboeuf) [inline tests/integration.rs]: Just a nit but you could factorize a little bit more like that:
```suggestion
        let net_params = if let Some(tap_str) = tap {
            format!(
                "tap={},ip={},mask=255.255.255.0,sock={},num_queues={},queue_size=1024",
                tap_str, ip, vunet_socket_path, num_queues
            )
        } else {
            format!(
                "ip={},mask=255.255.255.0,sock={},num_queues={},queue_size=1024",
                ip, vunet_socket_path, num_queues
            )
        }
        let child = Command::new("target/release/cloud-hypervisor")
            .args(&[
                "--net-backend",
                net_params.as_str(),
            ])
            .spawn()
            .unwrap();
```

File: tests/integration.rs [hunk 4]
```diff
@@ -1124,35 +1164,38 @@ mod tests {
     }
 
     #[cfg_attr(not(feature = "mmio"), test)]
-    fn test_vhost_user_net() {
+    fn test_vhost_user_net(
```
Comment (sboeuf) [inline tests/integration.rs]: Oops! You need to remove `#[cfg_attr(not(feature = "mmio"), test)]` to let the tests compile correctly.

**Comment by sboeuf** [comment]
@rbradford ready to be merged, PTAL

File: tests/integration.rs [hunk 5]
```diff
@@ -1208,6 +1241,28 @@ mod tests {
...
 
+    #[cfg_attr(not(feature = "mmio"), test)]
+    fn test_vhost_user_net_default() {
+        test_vhost_user_net(None, 2, &prepare_vhost_user_net_daemon)
+    }
+
+    #[cfg_attr(not(feature = "mmio"), test)]
+    fn test_vhost_user_net_tap() {
+        let vunet_tap0 = "vunet-tap0";
+        test_vhost_user_net(Some(vunet_tap0), 2, &prepare_vhost_user_net_daemon)
```
Comment (rbradford) [inline tests/integration.rs]: big nit: Why the local `vunet_tap0` local? It's not even the same name in the two tests that use it so you might as well just do `Some("vunet_tap0")`

File: tests/integration.rs [hunk 6]
```diff
@@ -1208,6 +1241,28 @@ mod tests {
...
+    }
+
+    #[cfg_attr(not(feature = "mmio"), test)]
+    fn test_vhost_user_net_multiple_queues() {
+        test_vhost_user_net(None, 4, &prepare_vhost_user_net_daemon)
+    }
+
+    #[cfg_attr(not(feature = "mmio"), test)]
+    fn test_vhost_user_net_tap_multiple_queues() {
+        let vunet_tap1 = "vunet-tap1";
```
Comment (rbradford) [inline tests/integration.rs]: Ditto.

File: tests/integration.rs (hunk 5 above)
Comment (sboeuf) [reply tests/integration.rs]: You mean `Some("vunet-tap0")`, right?

**Pull request merged**